Rearrange page search

sascha-karnatz opened the pull request:
Previously the page and the ingredients (and in previous versions also essences) were stored as separate PgSearch::Document entries. On search these document were combined to a single document to make it usable in the search itself. These mechanic made it pretty difficult to extend the search with other models. In newer versions of Alchemy the page is only available after a new page version is released and these mechanic is used to create a single PgSearch::Document with the content of the page and all supported ingredients. This change makes the whole search and the index creation less complex.

sascha-karnatz committed:
Store Alchemy::Page as a single PgSearch::Document

Previously the page and the ingredients (and in previous versions also essences) were stored as separate PgSearch::Document entries. On search these document were combined to a single document to make it usable in the search itself. These mechanic made it pretty difficult to extend the search with other models. In newer versions of Alchemy the page is only available after a new page version is released and these mechanic is used to create a single PgSearch::Document with the content of the page and all supported ingredients. This change makes the whole search and the index creation less complex.

The Alchemy module now has searchable_ingredients attribute which makes it easier to change the indexing from the outside. Furthermore a few extensions are now in a Search namespace, because they are not related to the PgSearch gem.

tvdeyen commented on app/extensions/alchemy/pg_search/ingredient_extension.rb:
```diff
@@ -1,18 +1,15 @@
 module Alchemy::PgSearch::IngredientExtension
+  def searchable_content
+    content_field = case type
+      when "Alchemy::Ingredients::Richtext"
```
I think it makes sense to make this a Map as a replacement for `SEARCHABLE_INGREDIENTS`

```rb
Alchemy.searchable_ingredients = {
  "Alchemy::Ingredients::Richtext" => :stripped_body,
  "Alchemy::Ingredients::Picture" => :caption
}
```

tvdeyen commented on app/extensions/alchemy/pg_search/ingredient_extension.rb:
```diff
@@ -1,18 +1,15 @@
+    send(content_field)
   end
 
   def searchable?
```
We can remove `Alchemy::PgSearch.is_searchable?(type)`

tvdeyen commented on app/extensions/alchemy/pg_search/ingredient_extension.rb:
```diff
@@ -1,18 +1,15 @@
 module Alchemy::PgSearch::IngredientExtension
```
This module could be a more generic `Alchemy::Search::Searchable` module

tvdeyen commented on app/extensions/alchemy/pg_search/ingredient_extension.rb:
```diff
@@ -24,7 +21,3 @@ def searchable?
 
 # add the PgSearch model to all ingredients
 Alchemy::Ingredient.prepend(Alchemy::PgSearch::IngredientExtension)
```
```rb
Alchemy::Ingredient.prepend(Alchemy::Search::Searchable)
```

tvdeyen commented on lib/alchemy-pg_search.rb:
```diff
@@ -19,19 +19,17 @@ def self.is_searchable?(ingredient_type)
-      [Alchemy::Page, Alchemy::Ingredient].each do |model|
-        ::PgSearch::Multisearch.rebuild(model)
-      end
+      Alchemy::Page.all.each{ |page| index_page(page) }
```
we should wrap this in a transaction

tvdeyen commented on lib/alchemy-pg_search.rb:
```diff
@@ -41,12 +39,16 @@ def self.remove_page(page)
+          document.content += " #{ingredient.searchable_content}"
         end
       end
+      document.content.strip!
```
```rb
document.content.squish!
```

tvdeyen commented on app/extensions/alchemy/pg_search/ingredient_extension.rb:
```diff
@@ -1,18 +1,15 @@
+        :caption
+      else
+        :value
+      end
```
We should add `squish` here

tvdeyen commented on lib/alchemy-pg_search.rb:
```diff
@@ -41,12 +39,16 @@ def self.remove_page(page)
       page.all_elements.includes(:ingredients).find_each do |element|
-        element.ingredients.select { |i| Alchemy::PgSearch.is_searchable?(i.type) }.each do |ingredient|
-          ingredient.update_pg_search_document
+        element.ingredients.select { |i| i.searchable? }.each do |ingredient|
```
We could use `map { _1.searchable_content }.join(" ")` here.

tvdeyen replied:
and maybe add `ingredients.searchable` scope

tvdeyen replied:
and add a `elements.searchable` scope as well.

tvdeyen commented on lib/alchemy-pg_search.rb:
```diff
@@ -6,7 +6,7 @@ module Alchemy
 
   module PgSearch
-    SEARCHABLE_INGREDIENTS = %w[Text Richtext Picture]
+    SEARCHABLE_INGREDIENTS = %w[Text Richtext Headline Picture]
```
👍🏻

sascha-karnatz committed:
Add support for the headline ingredient

Add headline to the list of supported ingredients. It contains valuable information and should be in the search index.

tvdeyen commented on lib/alchemy-pg_search.rb:
```diff
@@ -41,12 +39,15 @@ def self.remove_page(page)
+      document = page.update_pg_search_document
+      return if document.nil?
+      
       page.all_elements.includes(:ingredients).find_each do |element|
```
Maybe we have an N+1 here

```rb
page.all_elements.includes(ingredients: {element: {page_version: :page}}).
```

sascha-karnatz replied:
changed it and it is working without n+1s

tvdeyen commented on lib/alchemy-pg_search.rb:
```diff
@@ -57,12 +58,16 @@ def self.index_page(page)
+        # odd left join string
+        # Ref: https://guides.rubyonrails.org/active_record_querying.html#using-a-string-sql-fragment
+        query = query
+                  .joins("LEFT JOIN alchemy_pages ON alchemy_pages.id = pg_search_documents.page_id")
```
Do we need the `left_join` outside of the `if` statement?

sascha-karnatz replied:
It would make the query slower and Rails would reload the page again.

tvdeyen replied:
So, you are saying with a `ability` passed in we get N+1s?

tvdeyen commented on lib/alchemy-pg_search.rb:
```diff
@@ -5,24 +5,23 @@ module Alchemy
-      [Alchemy::Page, Alchemy::Ingredient].each do |model|
-        ::PgSearch::Multisearch.rebuild(model)
+      ActiveRecord::Base.transaction do
+        ::PgSearch::Document.delete_by(searchable_type: "Alchemy::Page")
```
```suggestion
        ::PgSearch::Document.delete_all
```

tvdeyen commented on lib/alchemy-pg_search.rb:
```diff
@@ -41,12 +40,15 @@ def self.remove_page(page)
     def self.index_page(page)
       remove_page page
```
We do not need this anymore

```suggestion
```

tvdeyen commented on lib/alchemy-pg_search.rb:
```diff
@@ -39,14 +38,16 @@ def self.remove_page(page)
+        element.ingredients.select { |i| i.searchable? }.map(&:searchable_content).join(" ")
+      end.join(" ")
+
+      document.update(content: "#{document.content} #{ingredient_content}".squish)
```
Maybe we need to `update_columns`

```suggestion
      document.update_columns(content: "#{document.content} #{ingredient_content}".squish)
```
